tempto-kafka/src/main/java/io/prestodb/tempto/fulfillment/table/kafka/KafkaTableManager.java

```
@@ -0,0 +1,225 @@
+        props.put("acks", "all");
+        props.put("retries", 0);
+        props.put("key.serializer", "org.apache.kafka.common.serialization.ByteArraySerializer");
+        props.put("value.serializer", "org.apache.kafka.common.serialization.ByteArraySerializer");
+
+        Producer<byte[], byte[]> producer = new KafkaProducer<>(props);
+
+        Iterator<KafkaMessage> messages = dataSource.getMessages();
+        while (messages.hasNext()) {
+            KafkaMessage message = messages.next();
+            try {
+                producer.send(new ProducerRecord<>(topic,
+                        message.getPartition().isPresent() ? message.getPartition().getAsInt() : null,
+                        message.getKey().orElse(null),
+                        message.getValue())).get();
+            }
+            catch (Exception e) {
+                throw new RuntimeException("could not send message to topic " + topic);
+            }
+        }
+    }
+
+    private void withZookeeper(Consumer<ZkUtils> routine)
+    {
+        int sessionTimeOutInMs = 15_000;
+        int connectionTimeOutInMs = 10_000;
+        String zookeeperHosts = zookeeperHost + ":" + zookeeperPort;
+
+        ZkClient zkClient = new ZkClient(zookeeperHosts,
+                sessionTimeOutInMs,
+                connectionTimeOutInMs,
+                ZKStringSerializer$.MODULE$);
+        try {
+            ZkUtils zkUtils = new ZkUtils(zkClient, new ZkConnection(zookeeperHosts), false);
+            routine.accept(zkUtils);
+        }
+        finally {
+            zkClient.close();
+        }
+    }
+
+    @Override
+    public TableInstance<KafkaTableDefinition> createMutable(KafkaTableDefinition tableDefinition, MutableTableRequirement.State state, TableHandle tableHandle)
+    {
+        throw new IllegalArgumentException("Mutable tables are not supported by KafkaTableManager");
+    }
+
+    @Override
+    public void dropTable(TableName tableName)
+    {
+        // noop
```

Why not `throw Unsupported`?

---

We could apparently. Changed.